src/reference-manual/compiler-optimizations.Rmd

```diff
@@ -0,0 +1,487 @@
+
+The levels include these optimizations:
+
+-   **O0** includes no optimizations.
+-   **O1** includes:
+    -   Dead code elimination
+    -   Auto-differentiation level optimization
+-   **Oexperimental** includes optimizations specified by **O1** and also:
+    -   One step loop unrolling
+    -   Constant propagation
+    -   Expression propagation
+    -   Copy propagation
+    -   Partial evaluation
+    -   Lazy code motion
+    -   Function inlining
+    -   Static loop unrolling
+
+In addition, **Oexperimental** will apply more repetitions of the optimizations, which may increase compile times.
+
+## Optimization descriptions
+
+
+### **O1** Optimizations
+
+
+#### Dead code elimination
+
+Dead code is code that does not have any effect on the behavior of the program. Code is not dead if it affects `target`, the value of any outside-observable variable like transformed parameters or generated quantities, or side effects such as print statements. Removing dead code can speed up a program by avoiding unnecessary computations.
+
+Example Stan program:
+
+```
+model {
+  int i;
+  i = 5;
+  for (j in 1:10);
+  if (0) {
+    print("Dead code");
+  } else {
+    print("Hi!");
+  }
+}
+```
+
+Compiler representation of program **before dead code elimination** (simplified from the output of `--debug-transformed-mir-pretty`):
+
+```
+log_prob {
+  int i = 5;
+  for(j in 1:10) {
+    ;
+  }
+  if(0) {
+    FnPrint__("Dead code");
+  } else {
+    FnPrint__("Hi!");
+  }
+}
+```
+
+Compiler representation of program **after dead code elimination** (simplified from the output of `--debug-optimized-mir-pretty`):
+
+```
+log_prob {
+  int i;
+  FnPrint__("Hi!");
+}
+```
+
+#### Constant propagation
+
+Constant propagation replaces uses of a variable which is known to have a constant value `C` with that constant `C`. This removes the overhead of looking up the variable, and also makes many other optimizations possible (such as static loop unrolling and partial evaluation).
+
+Example Stan program:
+
+```
+transformed data {
+  int n = 100;
+  int a[n];
+  for (i in 1:n) {
+    a[i] = i;
+  }
+}
+```
+
+Compiler representation of program **before constant propagation** (simplified from the output of `--debug-transformed-mir-pretty`):
+
+```
+prepare_data {
+  data int n = 100;
+  data array[int, n] a;
+  for(i in 1:n) {
+    a[i] = i;
+  }
+}
+```
+
+Compiler representation of program **after constant propagation** (simplified from the output of `--debug-optimized-mir-pretty`):
+
+```
+prepare_data {
+  data int n = 100;
+  data array[int, 100] a;
+  for(i in 1:100) {
+    a[i] = i;
+  }
+}
+```
+
+#### Copy propagation
+
+Copy propagation is similar to [expression propagation](#org4e2cadd), but only propagates variables rather than arbitrary expressions. This can reduce the complexity of the code for other optimizations such as expression propagation.
```

Comment:
```suggestion
Copy propagation is similar to [expression propagation](#org4e2cadd) but only propagates variables rather than arbitrary expressions. This optimization can reduce the complexity of the code for other optimizations such as expression propagation.
```